File: c++ implementations/event_listener/key_processor.cpp
#include "key_processor.h"
#include <iostream>
#include <algorithm>
#include <chrono>

KeyProcessor::KeyProcessor() {
    modifier_keys = {
        "ctrl_l", "ctrl_r", "alt_l", "alt_r", "shift_l", "shift_r",
        "cmd_l", "cmd_r", "tab", "esc", "f1", "f2", "f3", "f4", "f5",
        "f6", "f7", "f8", "f9", "f10", "f11", "f12"
    };
    control_keys = {
        "enter", "backspace", "delete", "home", "end", "page_up",
        "page_down", "up", "down", "left", "right", "insert"
    };
    ctrl_map = {
        {'\x01', 'a'}, {'\x02', 'b'}, {'\x03', 'c'}, {'\x04', 'd'},
        {'\x05', 'e'}, {'\x06', 'f'}, {'\x07', 'g'}, {'\x08', 'h'},
        {'\x09', 'i'}, {'\x0a', 'j'}, {'\x0b', 'k'}, {'\x0c', 'l'},
        {'\x0d', 'm'}, {'\x0e', 'n'}, {'\x0f', 'o'}, {'\x10', 'p'},
        {'\x11', 'q'}, {'\x12', 'r'}, {'\x13', 's'}, {'\x14', 't'},
        {'\x15', 'u'}, {'\x16', 'v'}, {'\x17', 'w'}, {'\x18', 'x'},
        {'\x19', 'y'}, {'\x1a', 'z'}
    };
}

KeyProcessor::~KeyProcessor() {
    // Destructor
}

// Public getters to access private members from Python
std::unordered_set<std::string> KeyProcessor::get_pressed_keys() const { return pressed_keys; }
std::unordered_set<std::string> KeyProcessor::get_sent_hotkeys() const { return sent_hotkeys; }
double KeyProcessor::get_last_hotkey_time() const { return last_hotkey_time; }
double KeyProcessor::get_last_key_time() const { return last_key_time; }
bool KeyProcessor::get_is_typing_mode() const { return is_typing_mode; }
double KeyProcessor::get_typing_timeout() const { return typing_timeout; }

void KeyProcessor::add_pressed_key(const Key& key) {
    if (pressed_keys.find(key.name) == pressed_keys.end()) {
        pressed_keys.insert(key.name);
        key_press_times[key.name] = static_cast<double>(std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::high_resolution_clock::now().time_since_epoch()).count()) / 1000000.0;
        last_key_time = key_press_times[key.name];
    }
}

void KeyProcessor::remove_released_key(const Key& key) {
    pressed_keys.erase(key.name);
    key_press_times.erase(key.name);
}

std::string KeyProcessor::get_key_string(const Key& key) {
    std::string key_name = key.name;
    if (key_name == "ctrl_l" || key_name == "ctrl_r") return "ctrl";
    if (key_name == "alt_l" || key_name == "alt_r") return "alt";
    if (key_name == "shift_l" || key_name == "shift_r") return "shift";
    if (key_name == "cmd_l" || key_name == "cmd_r") return "cmd";
    return key_name;
}

std::string KeyProcessor::format_hotkey_combination() {
    std::vector<std::string> modifiers;
    std::vector<std::string> regular_keys;

    for (const auto& key_str : pressed_keys) {
        if (modifier_keys.count(key_str)) {
            modifiers.push_back(get_key_string({key_str}));
        } else {
            regular_keys.push_back(key_str);
        }
    }
    
    std::unordered_set<std::string> unique_modifiers_set;
    std::vector<std::string> unique_modifiers_vec;
    std::vector<std::string> modifier_order = {"ctrl", "alt", "shift", "cmd"};

    for(const auto& mod : modifier_order) {
        if (std::find(modifiers.begin(), modifiers.end(), mod) != modifiers.end() && unique_modifiers_set.find(mod) == unique_modifiers_set.end()) {
            unique_modifiers_vec.push_back(mod);
            unique_modifiers_set.insert(mod);
        }
    }
    for (const auto& mod : modifiers) {
        if (unique_modifiers_set.find(mod) == unique_modifiers_set.end()) {
            unique_modifiers_vec.push_back(mod);
        }
    }

    std::sort(regular_keys.begin(), regular_keys.end());
    unique_modifiers_vec.insert(unique_modifiers_vec.end(), regular_keys.begin(), regular_keys.end());

    std::string result = "";
    for (size_t i = 0; i < unique_modifiers_vec.size(); ++i) {
        result += unique_modifiers_vec[i];
        if (i < unique_modifiers_vec.size() - 1) {
            result += "+";
        }
    }
    return result;
}

bool KeyProcessor::is_hotkey_combination() {
    if (pressed_keys.empty()) {
        return false;
    }
    if (pressed_keys.size() == 1) {
        const auto& key_str = *pressed_keys.begin();
        if (key_str == "enter") return true;
        if (key_str == "backspace") return false;
        if (modifier_keys.count(key_str)) return false;
        if (control_keys.count(key_str)) return true;
    }
    bool has_modifier = false;
    bool has_non_modifier = false;
    for (const auto& key_str : pressed_keys) {
        if (modifier_keys.count(key_str)) has_modifier = true;
        else has_non_modifier = true;
    }
    if (has_modifier && has_non_modifier && pressed_keys.size() > 1) return true;
    if (pressed_keys.size() > 1 && !has_modifier) return true;
    return false;
}

bool KeyProcessor::should_send_hotkey(const Key& pressed_key) {
    std::string key_str = get_key_string_internal(pressed_key);
    if (key_str == "enter") return true;
    
    if (pressed_keys.size() == 1 && modifier_keys.count(key_str)) return false;
    if (pressed_keys.size() == 1 && !modifier_keys.count(key_str)) return true;

    bool has_modifier = false;
    bool has_non_modifier = false;
    for (const auto& k : pressed_keys) {
        if (modifier_keys.count(k)) has_modifier = true;
        else has_non_modifier = true;
    }
    if (has_modifier && has_non_modifier) {
        if (!modifier_keys.count(key_str)) return true;
        else return true;
    }
    return false;
}

void KeyProcessor::add_typed_character(const Key& key) {
    if (key.name == "backspace") {
        if (!typed_characters.empty()) typed_characters.pop_back();
    } else if (key.name == "space") {
        typed_characters += ' ';
    } else {
        if (key.character != '\0' && key.character >= 32) typed_characters += key.character;
    }
}

std::string KeyProcessor::flush_typing() {
    std::string text = typed_characters;
    typed_characters.clear();
    return text;
}

void KeyProcessor::reset_state() {
    pressed_keys.clear();
    sent_hotkeys.clear();
    key_press_times.clear();
    typed_characters = "";
    is_typing_mode = false;
}

std::string KeyProcessor::get_key_string_internal(const Key& key) {
    if (key.character != '\0') {
        if (key.character >= 32) return std::string(1, key.character);
        else {
            auto it = ctrl_map.find(key.character);
            if (it != ctrl_map.end()) return std::string(1, it->second);
            return std::string(1, key.character);
        }
    } else {
        std::string key_name = key.name;
        if (key_name == "ctrl_l" || key_name == "ctrl_r") return "ctrl";
        if (key_name == "alt_l" || key_name == "alt_r") return "alt";
        if (key_name == "shift_l" || key_name == "shift_r") return "shift";
        if (key_name == "cmd_l" || key_name == "cmd_r") return "cmd";
        return key_name;
    }
}

File: training_assistant/utils/event_listener/key_processor.py
# key_processor.py
from pynput import keyboard
import logging

logger = logging.getLogger(__name__)

class KeyProcessor:
    """Handles key processing, hotkey detection, and typing analysis"""
    
    def __init__(self):
        self.pressed_keys = set()
        self.last_hotkey_time = 0
        self.hotkey_timeout = 0.1  # 100ms timeout for hotkey detection
        self.typing_timeout = 0.5  # 500ms timeout for typing detection
        self.last_key_time = 0
        self.is_typing_mode = False
        self.sent_hotkeys = set()  # Track sent hotkey combinations to avoid duplicates
        self.key_press_times = {}  # Track when each key was first pressed
        self.typed_characters = ""
        
        # Modifier keys that typically indicate hotkeys
        self.modifier_keys = {
            keyboard.Key.ctrl_l, keyboard.Key.ctrl_r,
            keyboard.Key.alt_l, keyboard.Key.alt_r, 
            keyboard.Key.shift_l, keyboard.Key.shift_r,
            keyboard.Key.cmd_l, keyboard.Key.cmd_r,  # For Mac
            keyboard.Key.tab, keyboard.Key.esc,
            keyboard.Key.f1, keyboard.Key.f2, keyboard.Key.f3, keyboard.Key.f4,
            keyboard.Key.f5, keyboard.Key.f6, keyboard.Key.f7, keyboard.Key.f8,
            keyboard.Key.f9, keyboard.Key.f10, keyboard.Key.f11, keyboard.Key.f12
        }
        
        # Keys that typically indicate navigation/control rather than typing
        self.control_keys = {
            keyboard.Key.enter, keyboard.Key.backspace, keyboard.Key.delete,
            keyboard.Key.home, keyboard.Key.end, keyboard.Key.page_up, 
            keyboard.Key.page_down, keyboard.Key.up, keyboard.Key.down,
            keyboard.Key.left, keyboard.Key.right, keyboard.Key.insert
        }
    
    def get_key_string(self, key):
        """Convert key object to string representation"""
        try:
            if hasattr(key, 'char') and key.char is not None:
                # Filter out control characters and hex values caused by Ctrl combinations
                char = key.char
                # If it's a control character (ASCII < 32), it's likely from Ctrl combination
                if ord(char) < 32:
                    # Common Ctrl combinations mapping
                    ctrl_map = {
                        '\x01': 'a',  # Ctrl+A
                        '\x02': 'b',  # Ctrl+B  
                        '\x03': 'c',  # Ctrl+C
                        '\x04': 'd',  # Ctrl+D
                        '\x05': 'e',  # Ctrl+E
                        '\x06': 'f',  # Ctrl+F
                        '\x07': 'g',  # Ctrl+G
                        '\x08': 'h',  # Ctrl+H (also backspace)
                        '\x09': 'i',  # Ctrl+I (also tab)
                        '\x0a': 'j',  # Ctrl+J (also enter/newline)
                        '\x0b': 'k',  # Ctrl+K
                        '\x0c': 'l',  # Ctrl+L
                        '\x0d': 'm',  # Ctrl+M (also enter)
                        '\x0e': 'n',  # Ctrl+N
                        '\x0f': 'o',  # Ctrl+O
                        '\x10': 'p',  # Ctrl+P
                        '\x11': 'q',  # Ctrl+Q
                        '\x12': 'r',  # Ctrl+R
                        '\x13': 's',  # Ctrl+S
                        '\x14': 't',  # Ctrl+T
                        '\x15': 'u',  # Ctrl+U
                        '\x16': 'v',  # Ctrl+V
                        '\x17': 'w',  # Ctrl+W
                        '\x18': 'x',  # Ctrl+X
                        '\x19': 'y',  # Ctrl+Y
                        '\x1a': 'z',  # Ctrl+Z
                        '\x1b': 'esc', # ESC
                    }
                    return ctrl_map.get(char, char)
                return char
            else:
                # Handle special keys with cleaner names
                key_name = str(key).replace('Key.', '')
                # Clean up common key names
                clean_names = {
                    'ctrl_l': 'ctrl',
                    'ctrl_r': 'ctrl',
                    'alt_l': 'alt',
                    'alt_r': 'alt',
                    'shift_l': 'shift',
                    'shift_r': 'shift',
                    'cmd_l': 'cmd',
                    'cmd_r': 'cmd'
                }
                return clean_names.get(key_name, key_name)
        except:
            return str(key)

    def format_hotkey_combination(self, keys):
        """Format the key combination into a readable string"""
        modifiers = []
        regular_keys = []
        
        for key in keys:
            key_str = self.get_key_string(key)
            if key in self.modifier_keys or key in self.control_keys:
                modifiers.append(key_str)
            else:
                regular_keys.append(key_str)
        
        # Sort modifiers for consistent output (remove duplicates for left/right variants)
        unique_modifiers = []
        seen_modifiers = set()
        
        modifier_order = ['ctrl', 'alt', 'shift', 'cmd']
        for mod in modifier_order:
            if mod in modifiers and mod not in seen_modifiers:
                unique_modifiers.append(mod)
                seen_modifiers.add(mod)
        
        # Add any other modifiers not in the order
        for mod in modifiers:
            if mod not in seen_modifiers:
                unique_modifiers.append(mod)
                seen_modifiers.add(mod)
        
        # Combine modifiers and regular keys
        key_strings = unique_modifiers + sorted(regular_keys)
        return '+'.join(key_strings)

    def is_hotkey_combination(self):
        """Determine if current key combination should be treated as a hotkey"""
        if len(self.pressed_keys) == 0:
            return False
            
        # Single key check
        if len(self.pressed_keys) == 1:
            key = next(iter(self.pressed_keys))
            
            # Enter is always a hotkey
            if key == keyboard.Key.enter:
                return True
                
            # Backspace is NOT a hotkey (used for typing correction)
            if key == keyboard.Key.backspace:
                return False
            
            # Modifier keys alone are NOT hotkeys (wait for combination)
            if key in self.modifier_keys:
                return False
            
            # Other control keys (except backspace) are hotkeys
            if key in self.control_keys and key != keyboard.Key.backspace:
                return True
        
        # Any combination with modifiers + other keys is a hotkey
        has_modifier = any(key in self.modifier_keys for key in self.pressed_keys)
        has_non_modifier = any(key not in self.modifier_keys for key in self.pressed_keys)
        
        if has_modifier and has_non_modifier and len(self.pressed_keys) > 1:
            return True
        
        # Multiple non-modifier keys pressed simultaneously are likely hotkeys
        if len(self.pressed_keys) > 1 and not has_modifier:
            return True
        
        return False
    
    def should_send_hotkey(self, pressed_key):
        """Determine if we should send the hotkey immediately"""
        # Always process Enter immediately
        if pressed_key == keyboard.Key.enter:
            return True
        
        # For modifier keys alone, don't send
        if len(self.pressed_keys) == 1 and pressed_key in self.modifier_keys:
            return False
        
        # For single non-modifier keys that are hotkeys (F1, arrows, etc.)
        if len(self.pressed_keys) == 1 and pressed_key not in self.modifier_keys:
            return True
        
        # For combinations: send when we have modifier + non-modifier and the non-modifier was just pressed
        has_modifier = any(k in self.modifier_keys for k in self.pressed_keys)
        has_non_modifier = any(k not in self.modifier_keys for k in self.pressed_keys)
        
        if has_modifier and has_non_modifier:
            # Check if a non-modifier key was just pressed (completing the combination)
            if pressed_key not in self.modifier_keys:
                return True
            # Or if this is the first modifier being added to an existing non-modifier
            elif pressed_key in self.modifier_keys and len([k for k in self.pressed_keys if k not in self.modifier_keys]) > 0:
                return True
        
        return False
    
    def add_typed_character(self, key):
        """Add character to typed buffer"""
        if key == keyboard.Key.backspace:
            self.typed_characters = self.typed_characters[:-1]
        elif key == keyboard.Key.space:
            self.typed_characters += ' '
        else:
            try:
                if hasattr(key, 'char') and key.char is not None:
                    # Don't add control characters to typed text
                    if ord(key.char) >= 32:
                        self.typed_characters += key.char
            except AttributeError:
                pass
    
    def flush_typing(self):
        """Return and clear typed characters"""
        if self.typed_characters:
            text = self.typed_characters
            self.typed_characters = ""
            return text
        return None
    
    def reset_state(self):
        """Reset all state variables"""
        self.pressed_keys.clear()
        self.sent_hotkeys.clear()
        self.key_press_times.clear()
        self.typed_characters = ""
        self.is_typing_mode = False